Sources/llbuild2/Core/Engine.swift

```diff
@@ -73,28 +77,59 @@ public protocol LLBFunction {
     func compute(key: LLBKey, _ fi: LLBFunctionInterface) -> LLBFuture<LLBValue>
 }
 
+public protocol LLBFunctionCache {
+    func get(key: LLBKey) -> LLBFuture<LLBDataID?>
+    func update(key: LLBKey, value: LLBDataID) -> LLBFuture<Void>
+}
+
 open class LLBTypedCachingFunction<K: LLBKey, V: LLBValue>: LLBFunction {
     public init() {}
 
+    private func computeAndUpdate(key: LLBKey, typedKey: K, _ fi: LLBFunctionInterface) -> LLBFuture<LLBValue> {
+        return self.compute(key: typedKey, fi).flatMap { (value: LLBValue) in
+            do {
+                return fi.db.put(try value.asCASObject()).flatMap { resultID in
+                    return fi.functionCache.update(key: key, value: resultID).map {
+                        return value
+                    }
+                }
+            } catch {
+                return fi.group.next().makeFailedFuture(error)
+            }
+        }
+    }
+
+    private func unpack<T>(_ object: LLBCASObject, _ fi: LLBFunctionInterface) throws -> T where T:  LLBPolymorphicSerializable {
+        return try T.init(from: object, registry: fi.registry)
+    }
+    private func unpack<T>(_ object: LLBCASObject, _ fi: LLBFunctionInterface) throws -> T where T: LLBCASObjectConstructable {
+        return try T.init(from: object)
+    }
+    private func unpack<T>(_ object: LLBCASObject, _ fi: LLBFunctionInterface) throws -> T {
+        fatalError("cannot unpack CAS object")
```

Comment:
I really want to have a `where V: LLBPolymorphicSerializable | LLBCASObjectConstructable` at the class level, but alas there is no way to express an `or` relationship.   And Swift won't let it compile unless all the possible types are covered.